### DISCUSSION TITLE
Indicadores fundamentalistas

### DISCUSSION THREAD

**Discussion by walterpitman** [dude333/rapina#31] [Ideas]
primeiramente parabens pelo codigo. 
eu tenho alguns indicadores que gosto de usar porem nao encontro eles normalmente nos sites habituais.

se tiver como,
como eu faço pra ajudar a colocar estes indicadores na planilha de excel?  e tambem o valuation da empresa?

obrigado

**Comment by dude333** [comment]
Opa, obrigado!

Dependendo de quais indicadores, talvez possam ser implementados dentro da função [metricsList](https://github.com/dude333/rapina/blob/66aeb609b4fdff2a2d1484f41b39748b40a91a07/reports/reports.go#L482).

Você tem alguns exemplos de quais indicadores você usa?

**Comment by walterpitman** [comment]
este site do damodaran mostra indicadores http://pages.stern.nyu.edu/~adamodar/New_Home_Page/datafile/variable.htm
bem interessante pra comecar dentre eles tem o ciclo financeiro da empresa 
Capital de giro (cdg):                                 Pass N circ + Patr Liq - Ativ N Circ
Necessidade de capital de giro  (ncg):    (Ativo Circulante - caixa ) - (Pass circulante - dividendos - emprestimo)
Saldo de tesouraria: CDG - NCG
sou zerado em programação mas queria ajudar de alguma forma.

outra coisa interessante era computar em planilha todos os indicadores das empresas de cada ano. tipo pra fazer um filtro comparativo. igual tem o site fundamentus, porém só tem la dados mais recentes.

o Capex vejo que nao aparece, pq o fluxo de caixa de investimento nao esta separado. so aparece o total. sabe como poderiamos pegar esta informação?
neste site https://www.rad.cvm.gov.br/ENET/frmGerenciaPaginaFRE.aspx?NumeroSequencialDocumento=80947&CodigoTipoInstituicao=1
aparece detalhado o fluxo de caixa.

**Comment by walterpitman** [comment]
Bom dia!
Se quiser também me passar a literatura da programaçao para eu entender. ai posso colaborar se precisar!

abraço

**Comment by dude333** [comment]
Opa, seria uma boa, pois estou sem tempo ultimamente. Eu até estava vendo de como usar a evolução das contas no modelo dinâmico do fleuriet, quem sabe no futuro :)

Eu usei a linguagem Go, você pode começar a estudar por aqui:

https://go-tour-br.appspot.com

Abs.

**Comment by dahn94** [comment]
O modelo do Fleuriet seria muito interessante. Parabéns pelo projeto.
@dude333

**Comment by dude333** [comment]
Oi @walterpitman, eu incluí alguns indicadores extras mais os que você passou (que acredito serem o modelo de Fleuriet) no commit c80ed4d69fee47e9fc6b1342a507598affb8b337.

Eu tentei achar, sem sucesso, os detalhes destas fórmulas que você passou:

```
Capital de giro (cdg): Pass N circ + Patr Liq - Ativ N Circ
Necessidade de capital de giro (ncg): (Ativo Circulante - caixa ) - (Pass circulante - dividendos - emprestimo)
Saldo de tesouraria: CDG - NCG
```

Geralmente o capital de giro é calculado subtraindo o PassCirc do AtivoCirc, mas eu deixei no relatório a fórmula que você passou.

Seguem alguns exemplos e agradeço se puder conferi-los.

### Engie:


  | [2009] | [2010] | [2011] | [2012] | [2013] | [2014] | [2015] | [2016] | [2017] | [2018] | [2019] | [2020]
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Capital de Giro (CDG) | 466 | (353) | 54 | 295 | 155 | 1.331 | 1.454 | 1.554 | (1.941) | 386 | 766 | 2.352
Necessidade de Capital de Giro   (NCG) | 3.056 | 3.453 | 4.096 | 3.919 | 3.656 | 4.331 | 3.825 | 3.883 | 4.461 | 8.914 | 13.502 | 16.419
Saldo de Tesouraria (T) | (2.590) | (3.806) | (4.043) | (3.625) | (3.502) | (3.000) | (2.371) | (2.328) | (6.402) | (8.527) | (12.736) | (14.067)

### WEG:


  | [2009] | [2010] | [2011] | [2012] | [2013] | [2014] | [2015] | [2016] | [2017] | [2018] | [2019] | [2020]
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Capital de Giro (CDG) | 2.275 | 2.855 | 3.114 | 2.697 | 4.274 | 4.717 | 6.094 | 5.849 | 5.089 | 4.405 | 5.270 | 6.674
Necessidade de Capital de Giro   (NCG) | 4.314 | 5.388 | 6.659 | 3.270 | 4.293 | 5.774 | 8.235 | 7.020 | 6.201 | 6.318 | 6.148 | 5.467
Saldo de Tesouraria (T) | (2.040) | (2.533) | (3.545) | (573) | (19) | (1.057) | (2.140) | (1.171) | (1.112) | (1.913) | (878) | 1.207

**Comment by walterpitman** [comment]
oi @dude333 
"Geralmente o capital de giro é calculado subtraindo o PassCirc do AtivoCirc, mas eu deixei no relatório a fórmula que você passou"
sim! Ativo circulante - passivo circulante irá dar o mesmo resultado. pode fazer nele se ficar mais fácil.

fiz um report da weg no app rapina e nao achei as métricas Fleuriet

mas vou te colocar aqui o número das contas do balanço patrimonial

**Saldo de tesouraria: 
(1.01.01 | Caixa e Equivalentes de Caixa + 1.01.02 | Aplicações   Financeiras ) - ( 2.01.04 | Empréstimos e Financiamentos + 2.01.05.02.01 | Dividendos e JCP   a Pagar + 2.01.05.02.02 | Dividendo Mínimo   Obrigatório a Pagar )

Capital de giro:
(1.01 | Ativo Circulante) - (2.01 | Passivo Circulante)**

**NCG: CG - ST**

Acho que dessa forma fica mais fácil


Me diga o que acha!!? grande abraço

**Comment by walterpitman** [comment]
segue exemplo da Weg


ST     | 1.194 | 1.471 | 1.227 | 838 | 2.373 | 2.615 | 2.978 | 3.581 | 2.398 | 1.315 | 2.309 | 3.707
CDG | 2.275 | 2.855 | 3.114 | 2.697 | 4.274 | 4.717 | 6.094 | 5.849 | 5.089 | 4.405 | 5.270 | 6.674
NCG | 1.080 | 1.385 | 1.887 | 1.859 | 1.900 | 2.103 | 3.117 | 2.267 | 2.691 | 3.089 | 2.961 | 2.967

Mostrando inclusive muito mais saúde financeira
acho que com as métricas que te passei.

os cálculos ficarão tops!!!

**Reply by dude333** [reply]
@walterpitman, veja o último release: https://github.com/dude333/rapina/releases/tag/v0.13.1

Para mostrar os novos indicadores você precisa usar o flag `-all`:

    ./rapina report -all weg

**Reply by walterpitman** [reply]
Acabei de testar!! Está Perfeito!
e o arquivo da CVM está realmente mais completo.

o projeto está cada vez mais sensacional

se me permite... mais algumas sugestões

CAPEX: 6.02.02 | Imobilizado + 6.02.03 | Intangível

ABRAÇOS

**Reply by dude333** [reply]
Ótimo, vou adicionar aqui.

Só um detalhe, eu passei a calcular o ROE usando o Patrimônio Líquido médio, ou seja, (PL_atual + PL_ano.anterior)/2.

**Reply by dude333** [reply]
Essa parte do Imobilizado não está muito bem padronizado, vou ver se tenho alguma ideia aqui:

    6.02.01 Aquisição de ativo imobilizado
    6.02.01 Aquisição de ativo imobilizado e intangível
    6.02.01 Aquisição de imobilizado
    6.02.01 Imobilizado
    6.02.02 Adições ao imobilizado
    ...

**Reply by walterpitman** [reply]
Oi Dude, 

não sei se já descobriu como pegar o capex... mas olhando de novo as métricas que me passou... são formas diferentes de descrever a mesma coisa. talvez mude de CIA para CIA mas são a mesma coisa.

acho que somar o 6.02.01 + 6.02.02... independente do que vier escrito tem chance de estar certo.
caso os nomes sejam esses mesmo. que se referem a mesma coisa

**Comment by dude333** [comment]
Opa, legal. Ainda vou conferir se esses itens aparecem em todas as empresas (geralmente o código acaba mudando em alguns casos).

Para mostrar as métricas de Fleuriet que eu fiz até o momento é necessário passar o flag `-F`. Ou pode passar o `--all` para mostrar todas as métricas:

    ./rapina report -all weg

Mas para isso você precisa compilar o último commit (a versão que está na aba de releases não tem essa função ainda).

Depois quero ver se faço um relatório com as métricas de todas as empresas.

**Reply by walterpitman** [reply]
Boa tarde Dude. Como eu compilo o último commit? é o arquivo master? não tem a pasta CLI nele.

**Reply by dude333** [reply]
Oi Walter, boa tarde! Eu mudei a estrutura dos diretórios. Agora para compilar o código você precisa estar no diretório principal do rapina e rodar o `make` (eu movi o Makefile para esta pasta). Ele vai compilar e gerar o código na pasta `bin`.

**Reply by dude333** [reply]
Eu incluí um servidor web para roda na sua própria máquina nessa última versão, mas por hora só um relatório: o de rendimentos de FIIs.

O executável dessa versão está aqui: https://github.com/dude333/rapina/releases/tag/v1.0.1

Exemplo:

<img src="https://imgur.com/Jgzu6Jb.jpg" height="720" />

**Reply by walterpitman** [reply]
Top demais!!!
Quando a CVM adicionar então! vai ficar top. até setembro no máximo eles devem ter o arquivo csv pronto.

vi ali que teve uma adição do yahoo? é pra pegar cotação?

**Reply by dude333** [reply]
Eu comecei com a www.alphavantage.co, mas aí descobri que você só consegue usar a API 5x por minuto. Então eu passei para o Yahoo, mas em alguns casos o histórico não está completo (acho que o da NSLU11 é um deles), então fui direto à fonte e estou pegando da B3 mesmo. Como o site vive mostrando a mensagem de sistema indisponível, achei que teria problema, mas não falhou nenhuma vez nos testes que eu fiz.

**Comment by walterpitman** [comment]
@dude333 concordo. Até me assustei. Parece que o arquivo da cvm ficou mais completo. O fluxo de caixa está totalmente dividido. Aparece até o capex agora. Vou ver se é pra todas também. Acho que é melhoria dos dados abertos mesmo. Tomara.

Vou tentar achar esse commit. Mas nem sei o que é. Kkkkkk. Tenho tanta ideia. Pena que não manjo de programação. 

Projeto espetacular este. Vamos pro próximo nível!

**Comment by walterpitman** [comment]
@dude333 uma pergunta.

vi que está adicionado o numero de shares por ano na empresa.

será que teriamos como ter o valor de mercado do fechamento do ano em 31/12 e junto com a maxima e minima daquele ano? e discriminado em cada ano?

porque daí dá pra calcular entrerprise value... e fora outras métricas que dependem disso. 
p/l , Ev/ebit e por ai vai...

obrigado

**Comment by dude333** [comment]
Estou vendo aqui um jeito de pegar o histórico de cotações direto da B3.

**Reply by PedroDnT** [reply]
Sei pear facil o historico mas pelo api do Yahoo... Na verdade tenho até uma base de dados com cotações, negociacoes, enfim, tudo da B3. Vou recuperar. Só sei python mas vou dar uma estudada em GO.

**Reply by dude333** [reply]
Oi @PedroDnT, eu consegui achar os arquivos da B3 :) E também do Yahoo e da Alpha Vantage. Eu faço uma busca nessa ordem (começando pela B3), pois já vi alguns casos que o Yahoo não tem o histórico completo. E a Alpha Vantage seria um bom recurso, mas é limitado em 5 consultas por minuto.

Esse aqui é o módulo com as funções para busca nestes 3 sites:

https://github.com/dude333/rapina/blob/d80b2bd0ca8e88b8a67ad1c793814f096b8e1481/fetch/fetch_stockquote.go

**Comment by bpenedo** [comment]
O modelo do Fleuriet foi sensacional, meus parabéns pela iniciativa. Outro modelo com a proposta realmente interessante é o Piotroski F-score => https://drive.google.com/drive/folders/1uFrz1wGS9LNKYqhizgFoCtcCt1NVI_7W?usp=sharing

Outros multiplos financeiros e modelos interessantes como Dupont tb são validos: https://marketxls.com/7BestValuationMethods-MarketXLS.pdf

[]s

**Reply by dude333** [reply]
Oi @bpenedo, o maior problema em montar os indicadores é que os dados publicados na CVM nem sempre são padronizados. A lógica principal dessa parte dos indicadores vem destas duas partes:

## Tabela com os itens dos relatórios financeiros (BP, DRE, FC...):

https://github.com/dude333/rapina/blob/d80b2bd0ca8e88b8a67ad1c793814f096b8e1481/parsers/codeaccounts.go#L70-L115

---

## Cálculo dos indicadores

Depois essa função é que efetivamente faz os cálculos com base na classificação da tabela acima:

https://github.com/dude333/rapina/blob/d80b2bd0ca8e88b8a67ad1c793814f096b8e1481/reports/reports.go#L472-L547

---

Uma ajuda que seria bem vinda é de achar um padrão para que eu possa colocar os itens de uma métrica na tabela `_accountsTable` e a fórmula com base nesses itens.

(A função safeDiv é apenas uma divisão, mas que retorna zero caso o denominador seja zero.)

Abs!
